psydac/feec/derivatives.py

```diff
@@ -86,6 +257,15 @@ def domain(self):
     @property
     def codomain(self):
         return self._codomain
+    
+    def __call__(self, u):
+        assert isinstance(u, FemField)
+        assert u.space == self.domain
+
+        coeffs = self.matrix.dot(u.coeffs)
```

Comment:
Good catch @campospinto, I agree with you

Reply:
Yes; I'd also agree, that sounds very reasonable. Are there any suggestions on how to rename the `matrix` parameter? We would also then need to think about maybe how to potentially rename `psydac.api.feec.DiscreteDerham.derivatives_as_matrices` 

In addition, we could also introduce some `tostencil` and/or `tokronstencil` method on the `DiffOperator` class.

Reply:
If you have a look at the base classes `LinearOperator` and `Matrices`, the only difference is that the latter requires some [additional methods](https://github.com/pyccel/psydac/blob/296d077a1b61042463dcae0ef7eaeea18c180cfa/psydac/linalg/basic.py#L143-L188), the most important being `toarray` and `tosparse`. Based on this distinction, your new class is really a `Matrix`, but with a special internal representation.

Reply:
In other words, if you can convert a matrix-free linear operator to a sparse matrix, that operator is a `Matrix` according to the current design of Psydac. Maybe the two classes (`LinearOperator` and `Matrix`) really need to be merged into one, and then we can create some more abstract classes to provide general operations: 

- `LinearOperator`
- `_SumLinearOperator`
- `_ScaledLinearOperator`
- `_ProductLinearOperator`
- etc...

Just like it is done in Scipy: https://github.com/scipy/scipy/blob/master/scipy/sparse/linalg/interface.py

Reply:
Ok; I can change the type of the soon-to-be-renamed `KroneckerDifferentialOperator` from `LinearOperator` to `Matrix`, with all the `__add__` methods etc. first converting it to a `KroneckerStencilMatrix` (or even `StencilMatrix`) and then doing the `__add__`. (`__iadd__` etc. will fail with `NotImplementedError`)

About the `LinearOperator`, and `Matrix`: Would it maybe make sense, if the former works on infinite-dimensional vector spaces (i.e. at least `domain` or `codomain` is infinite-dimensional)? Since we of course could compute `toarray` and `tosparse` for each `LinearOperator` which works on finite-dimensional vector spaces only (by sending each standard basis element through `dot`). ... Or something like that.

But yes, such `LinearOperator`-derivative classes would be nice... But shouldn't all of this be rather discussed in a separate issue?

Reply:
@campospinto, @davschneller: I am realizing that this topic needs a separate discussion, and then probably a dedicated PR. So I am going to accept David's PR as it is for now.

Reply:
Ok -- actually I was about making some specific suggestions in my initial question, and I realized that I was not able to do so :)